### PR TITLE
Jetpack connect selectors state

### PR DESCRIPTION
**Pull request by johnHackworth**
As @aduth notes [here](https://github.com/Automattic/wp-calypso/pull/6460#discussion_r80930230), we are using a state subtree in the Jetpack Connect selectors instead of using the global state tree as we specify in our data style guide. 

This PRs change the selectors to use the global state tree. There should be no changes in the UI at all

How to test
=========
1. Go to http://calypso.localhost:3000/jetpack/connect and make sure you can complete a jetpack site connection  (plan selection included)
2. That's it 

ping @aduth @tyxla

### CODE REVIEW

File: client/signup/jetpack-connect/plans.jsx [hunk 1]
```diff
@@ -58,12 +58,11 @@ const Plans = React.createClass( {
 	},
 
 	componentWillReceiveProps( props ) {
-		const selectedSite = props.sites.getSelectedSite();
```
Comment (tyxla) [inline client/signup/jetpack-connect/plans.jsx]: Is there any particular reason to use the old props in `componentWillReceiveProps` instead of the new ones (as it was before)?

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 2]
```diff
@@ -640,24 +636,27 @@ const JetpackConnectAuthorizeForm = React.createClass( {
 
 export default connect(
 	state => {
-		const site = state.jetpackConnect.jetpackConnectAuthorize &&
-				state.jetpackConnect.jetpackConnectAuthorize.queryObject &&
-				state.jetpackConnect.jetpackConnectAuthorize.queryObject.site
+		const queryObjectSite = state.jetpackConnect.jetpackConnectAuthorize &&
```
Comment (tyxla) [inline client/signup/jetpack-connect/authorize-form.jsx]: We should access the state tree just by selectors, and not directly. This should probably be done in another PR though - I can work on it.

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 3]
```diff
@@ -640,24 +636,27 @@ const JetpackConnectAuthorizeForm = React.createClass( {
...
+			state.jetpackConnect.jetpackConnectAuthorize.queryObject &&
+			state.jetpackConnect.jetpackConnectAuthorize.queryObject.site;
+
+		const site = queryObjectSite
 			? getSiteByUrl( state, state.jetpackConnect.jetpackConnectAuthorize.queryObject.site )
```
Comment (aduth) [inline client/signup/jetpack-connect/authorize-form.jsx]: It might be clearer too to rename to `siteId`. I was about to drop a comment about how we shouldn't store site objects other than in `state.sites.items`, then realized this _is_ the ID already 😆 

(Fine for another pull request)

File: client/signup/jetpack-connect/plans.jsx (hunk 1 above)
Comment (aduth) [inline client/signup/jetpack-connect/plans.jsx]: >Is there any particular reason to use the old props in `componentWillReceiveProps` instead of the new ones (as it was before)?

Seconded. You probably want the treat the `props` argument as the most up to date. You could use `componentDidUpdate` instead if you'd rather use `this.props` (since this calls to many other functions on the component).

File: client/state/jetpack-connect/selectors.js [hunk 4]
```diff
@@ -5,18 +5,20 @@ const isCalypsoStartedConnection = function( state, siteSlug ) {
 		return false;
 	}
 	const site = siteSlug.replace( /.*?:\/\//g, '' );
-	if ( state && state[ site ] ) {
+	const sessions = state.jetpackConnect.jetpackConnectSessions;
+
+	if ( sessions && sessions[ site ] ) {
```
Comment (aduth) [inline client/state/jetpack-connect/selectors.js]: Looking at the reducer, `sessions` will _always_ be an object (initial state empty object), so the first truthiness check here is unnecessary.

File: client/state/jetpack-connect/selectors.js [hunk 5]
```diff
@@ -25,15 +27,18 @@ const getFlowType = function( state, site ) {
...
  */
 const hasXmlrpcError = function( state ) {
+	const authorizeData = state.jetpackConnect.jetpackConnectAuthorize;
+
 	return (
-		state.authorizeError &&
-		state.authorizeError.message &&
-		state.authorizationCode &&
-		state.authorizeError.message.indexOf( 'error' ) > -1
+		authorizeData &&
```
Comment (aduth) [inline client/state/jetpack-connect/selectors.js]: `authorizeData` will always be at least an empty object because of the reducer's initial state.

Also, for checking the existence of nested properties, Lodash's [`_.has`](https://lodash.com/docs/4.16.2#has) or [`_.get`](https://lodash.com/docs/4.16.2#get) may come in handy:

```js
const hasXmlrpcError = function( state ) {
	const authorizeData = state.jetpackConnect.jetpackConnectAuthorize;

	return (
		authorizeData.authorizationCode &&
		includes( get( authorizeData, [ 'authorizeError', 'message' ] ), 'error' )
	);
};
```

File: client/signup/jetpack-connect/authorize-form.jsx (hunk 3 above)
Comment (johnHackworth) [inline client/signup/jetpack-connect/authorize-form.jsx]: Hmm, that rename may not be that easy ... that queryObject is pretty much what Jetpack sends to calypso, so renaming it is out of reach (until a new jetpack release). We could of course just normalize the input and rename the attribute, but I'm not sure what's our common position on that (using the same attribute names than the API or not). I remember there was a discussion some time ago, but I don't remember getting to an agreement ... what do you think about it, @aduth  ?

File: client/signup/jetpack-connect/plans.jsx (hunk 1 above)
Comment (johnHackworth) [inline client/signup/jetpack-connect/plans.jsx]: oops, no... to be honest, this has been a case of a lousy search/replace 'selectedSite' by 'this.props.selectedSite'.

Anyway, now you mention it, @aduth , there's no reason to do all this in componentWillReceiveProps ... I could very well move it to ComponentDidUpdate and probably will be less confusing (as in nobody asking "why do you need to check all this before the component props are updated?")

File: client/signup/jetpack-connect/authorize-form.jsx (hunk 3 above)
Comment (johnHackworth) [inline client/signup/jetpack-connect/authorize-form.jsx]: oh, and for having this in a separate selector: yeah, that sounds good ... We can do that in other PR, as this one's scope was to update our current selectors

File: client/state/jetpack-connect/selectors.js (hunk 5 above)
Comment (johnHackworth) [inline client/state/jetpack-connect/selectors.js]: This is just a matter of personal preference, but I find `_.get` syntax way more confusing that chaining a couple of truthiness checks... I guess that may change if/when I get used to it, but anyway I think it's way more 'js canonical' to just use an extra check ...

Even if it's a line longer, I find 
```
return (
	authorizeData.authorizeError.message &&
	authorizeData.authorizationCode &&
	authorizeData.authorizeError.message.indexOf( 'error' ) > -1
);
```
more readable than
``` 
return (
        authorizeData.authorizationCode &&
        includes( get( authorizeData, [ 'authorizeError', 'message' ] ), 'error' )
    );
```
The first one reads like prose, the second one reads like math :)

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 6]
```diff
@@ -640,24 +636,27 @@ const JetpackConnectAuthorizeForm = React.createClass( {
...
+
 		return {
 			jetpackConnectAuthorize: state.jetpackConnect.jetpackConnectAuthorize,
 			jetpackSSOSessions: state.jetpackConnect.jetpackSSOSessions,
-			jetpackConnectSessions: state.jetpackConnect.jetpackConnectSessions,
 			isAlreadyOnSitesList: !! site,
 			isFetchingSites,
-			requestHasXmlrpcError
+			requestHasXmlrpcError,
+			calypsoStartedConnection: isCalypsoStartedConnection( state, queryObjectSite )
```
Comment (johnHackworth) [inline client/signup/jetpack-connect/authorize-form.jsx]: Hmm, we can't do that, because we are also using this selector in other parts of the code that passes a different parameter (also a site slug, but obtained from a different subtree).  Maybe we should make this an abstract selector and create a couple of new ones that deal with each case, but I'm not sure about if it would be adding more complexity for the sake of purity

**Comment by johnHackworth** [comment]
Ok, I updated some of the things @tyxla and @aduth  commented. Some of the improvements can be done in a follow-up PR

File: client/signup/jetpack-connect/authorize-form.jsx (hunk 3 above)
Comment (aduth) [inline client/signup/jetpack-connect/authorize-form.jsx]: >that queryObject is pretty much what Jetpack sends to calypso, so renaming it is out of reach (until a new jetpack release).

Ah, I didn't realize this. Well, another alternative to changing the structure of state is to use more selectors, so in this case `getSiteByUrl( state, getJetpackConnectAuthorizeSiteId( state ) );` which goes to the point of both making it more obvious that it's a site ID and also making state "opaque" (i.e. allowing us to change the structure later without too much trouble if we decide to).

Comment (tyxla) [inline client/signup/jetpack-connect/authorize-form.jsx]: Indeed @aduth - we intend to add some more selectors that use the full state tree, but let's do it in another PR and keep this one smaller.

**Comment by tyxla** [comment]
This LGTM, we can ship it unless @aduth has any further thoughts.

File: client/signup/jetpack-connect/authorize-form.jsx [hunk 7]
```diff
@@ -640,24 +636,27 @@ const JetpackConnectAuthorizeForm = React.createClass( {
...
-				state.jetpackConnect.jetpackConnectAuthorize.queryObject &&
-				state.jetpackConnect.jetpackConnectAuthorize.queryObject.site
+		const queryObjectSite = state.jetpackConnect.jetpackConnectAuthorize &&
+			state.jetpackConnect.jetpackConnectAuthorize.queryObject &&
+			state.jetpackConnect.jetpackConnectAuthorize.queryObject.site;
+
+		const site = queryObjectSite
 			? getSiteByUrl( state, state.jetpackConnect.jetpackConnectAuthorize.queryObject.site )
 			: null;
 		const requestHasXmlrpcError = () => {
```
Comment (aduth) [inline client/signup/jetpack-connect/authorize-form.jsx]: I'm confused why this is passed as a function and not just a boolean value. Fine to consider for a separate pull request.
